homeassistant/components/apple_tv/config_flow.py

```diff
@@ -0,0 +1,403 @@
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception")
+            abort_reason = "unrecoverable_error"
+
+        if abort_reason:
+            if self.pairing:
+                await self.pairing.close()
+            return self.async_abort(reason=abort_reason)
+
+        # Choose step depending on if PIN is required from user or not
+        if self.pairing.device_provides_pin:
+            return await self.async_step_pair_with_pin()
+
+        return await self.async_step_pair_no_pin()
+
+    async def async_step_pair_with_pin(self, user_input=None):
+        """Handle pairing step where a PIN is required from the user."""
+        errors = {}
+        if user_input is not None:
+            try:
+                self.pairing.pin(user_input[CONF_PIN])
+                await self.pairing.finish()
+                self.credentials[self.protocol.value] = self.pairing.service.credentials
+                return await self.async_begin_pairing()
+            except exceptions.PairingError:
+                _LOGGER.exception("Authentication problem")
+                errors["base"] = "auth"
+            except AbortFlow:
+                raise
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+        return self.async_show_form(
+            step_id="pair_with_pin",
+            data_schema=INPUT_PIN_SCHEMA,
+            errors=errors,
+            description_placeholders={"protocol": protocol_str(self.protocol)},
+        )
+
+    async def async_step_pair_no_pin(self, user_input=None):
+        """Handle step where user has to enter a PIN on the device."""
+        if user_input is not None:
+            await self.pairing.finish()
+            if self.pairing.has_paired:
+                self.credentials[self.protocol.value] = self.pairing.service.credentials
+                return await self.async_begin_pairing()
+
+            await self.pairing.close()
+            return self.async_abort(reason="device_did_not_pair")
+
+        pin = randrange(1000, stop=10000)
+        self.pairing.pin(pin)
+        return self.async_show_form(
+            step_id="pair_no_pin",
+            description_placeholders={
+                "protocol": protocol_str(self.protocol),
+                "pin": pin,
+            },
+        )
+
+    async def async_step_service_problem(self, user_input=None):
+        """Inform user that a service will not be added."""
+        if user_input is not None:
+            self.credentials[self.protocol.value] = None
+            return await self.async_begin_pairing()
+
+        return self.async_show_form(
+            step_id="service_problem",
+            description_placeholders={"protocol": protocol_str(self.protocol)},
+        )
+
+    async def _async_get_entry(self, protocol, name, credentials, address):
```

Comment:
We don't await inside so this doesn't need to be a coroutine.
```suggestion
    def _async_get_entry(self, protocol, name, credentials, address):
```